# Exercise Session 5 – Input/Output Analysis
### ENV–501 Material and Energy Flow Analysis

October 27th, 2022

Exercises readapted for Jupyter Notebook by Matteo Barsanti

In [1]:
# -- Import the required libraries
import numpy as np
import pandas as pd

## Exercise 1: Primary energy and CO2 emissions from automobile parts

The aim of this exercise is to evaluate the environmental impacts associated with one part of a car. The front panel is a static piece (see Figure 1), which is used to keep other parts of the car together. It does not have any other mechanical function. A supplier to the automotive industry faces a decision regarding the choice of material for this new structural part. The development department of the company has selected three materials that could satisfy the functional and production requirements. The three materials selected are steel, aluminum and a thermoset composite (polyester-based sheet molding compound). As a first step, study in details the case of aluminum.

To help with the decision making, this exercise evaluates the life cycle primary energy use and CO2 emissions of the aluminum front panel using a simplified Input-Output matrix. Its weight is 5.9 kg and it lasts for the average lifetime of an automobile or approximately 200’000 km. The amount of fuel specifically required to carry the front panel over this distance is 30.4 liters.

<ins>Input-Output and economic data</ins>

Input-Output tables are normally available at governmental bureau of statistics (for example the Bureau of Economy Analysis in the US). Table 1 presents a small extract of the monetary flows between the main sectors involved in the aluminum front panel production.
    
   
    
|                    | Aluminium    |  Coal \& Petroleum | Electricity | Total Outputs [M\\$] | 
| ---                |    :-:       |         :-:        |      :-:    |         :-:          | 
| Aluminium          |     976      |         0          |      0      |              5'688   | 
| Coal \& Petroleum  |     0.50     |         5'877      |      13'240 |              109'680 | 
| Electricity        |     1'518    |         1'243      |      27     |              132'400 | 

<center>Table 1: Transaction matrix [M\$] and total output [M\$]</center>


<ins>Questions:</ins>
1. Interpret the term 13’240 of the transaction matrix (table 1). What are the most important suppliers and customers for the aluminum sector (in monetary terms)?
2. Given the transaction matrix and the total output, complete the matrix of direct coefficients (A) and give an interpretation of its bottom row.

<ins>Environmental data</ins>

| Sector             | Energy Consumption [MJ/y]|  CO2 emissions [kg/y] | 
| ---                |    :-:                   |         :-:           |     
| Aluminium          |     0                    |         1.1E9         |       
| Coal \& Petroleum  |     6.26E13              |         76E9          |     
| Electricity        |     0                    |         1.5E12        | 

<center>Table 2: Total primary non-renewable energy consumption and CO2 emission per sector per year</center>


3. Using tables 1 and 2, calculate the environmental vectors for the primary energy consumption and for the CO2 emissions.

<ins>Case specific data:</ins>

4. To be able to evaluate the embodied energy and the CO2 emissions, case specific data has to be collected. Calculate the vector of final demand below.

| Goods             | Required amounts         |  Price                | Final demand  |
| ---                |    :-:                   |         :-:           |                |
| Aluminium          |     5.9 [kg/panel]       |         2.5 [\$/kg]   |                |       
| Oil for manufacturing  |     2.14 [l/panel]       |         0.32 [\$/l]   |                |     
| Electricity for manufacturing        |     15.2 [kWh/panel]     |         0.07 [\$/kWh] |                | 
| Gasoline during the use phase        |     30.4 [l/panel]       |         0.36 [\$/l]   |                | 

<center>Table 3: Aluminum front panel data and related prices</center>

5. Using the data provided in table 3, determine the necessary final output (in monetary units) in
each sector in order to manufacture one front panel. Identify the sector that generates the
most important contribution to the final output of the electricity sector.

6. Estimate the total non-renewable primary energy consumption induced by the manufacturing of a front panel.

<center><img src="ex1_car_front_panel.jpg" width="398"/></center>

<center>Figure 1: The front panel is the green part


Terminology adopted in the exercise: https://pymrio.readthedocs.io/en/latest/terminology.html 

#### EX1.1
*Interpret the term 13’240 of the transaction matrix (table 1). What are the most important suppliers and customers for the aluminum sector (in monetary terms)?*

In [2]:
# -- Data

# transation matrix
Z = np.array([[ 976,    0,     0,   5688], 
              [ 0.5, 5877, 13240, 109680], 
              [1518, 1243,    27, 132400]])
Z_df = pd.DataFrame(Z,
                    index   = ['Aluminium', 'Coal & Petroleum', 'Electricity'],
                    columns = ['Aluminium', 'Coal & Petroleum', 'Electricity', 'Total output xj'])
Z_df

Aluminium  Coal & Petroleum  Electricity  Total output xj
Aluminium             976.0               0.0          0.0           5688.0
Coal & Petroleum        0.5            5877.0      13240.0         109680.0
Electricity          1518.0            1243.0         27.0         132400.0

In the transation matrix $Z = z_{i,j}$, the cell 13’240 M\\$ corresponds to the output of $i$ the coal & petroleum sector to $j$ the electricity sector, and thus to the intermediate product provided by the coal & petroleum sector and consumed by the electricity sector. Destination of production of each sector: cells of the corresponding row.

The most important contributor to the aluminium sector is the electricity sector, which contributes with 1’518 M$ to the total demand of the aluminium sector. Demand of each sector: cells of the corresponding column.



#### EX1.2
*Given the transaction matrix and the total output, complete the matrix of direct coefficients (A) and give an interpretation of its bottom row.*

In [3]:
# -- Data

# array of total outputs xj
xj = Z_df.loc[:,'Total output xj'].copy()

# matrix of direct coefficients A = (aij) = (zij / xj) 
A = Z_df.loc[Z_df.index,Z_df.index].copy()
for i in A.index:
    for j in A.columns:
        A.loc[i,j] = Z_df.loc[i,j] / xj.loc[j]
A    

Aluminium  Coal & Petroleum  Electricity
Aluminium          0.171589          0.000000     0.000000
Coal & Petroleum   0.000088          0.053583     0.100000
Electricity        0.266878          0.011333     0.000204

These numbers correspond to the relative contribution of each intermediate product to the consuming sector’s final output to the economy.

Consequently, the bottom row corresponds to the relative contribution of the electricity sector to each of the other sectors’ final production. 

#### EX1.3
*Using tables 1 and 2, calculate the environmental vectors for the primary energy consumption and for the CO2 emissions.*

In [4]:
# -- Data

# environmental data matrix
E = np.array([[             0,    1.1 * 10**9], 
              [ 6.26 * 10**13,     76 * 10**9], 
              [             0,   1.5 * 10**12]])
E_df = pd.DataFrame(E,
                    index   = ['Aluminium', 'Coal & Petroleum', 'Electricity'],
                    columns = ['Energy consumption [MJ/y]', 'CO2 emissions [kg/y]'])
E_df

Energy consumption [MJ/y]  CO2 emissions [kg/y]
Aluminium                      0.000000e+00          1.100000e+09
Coal & Petroleum               6.260000e+13          7.600000e+10
Electricity                    0.000000e+00          1.500000e+12

In [5]:
# -- Calculation

# energy consumption per sector = E cons. j / Output j
# emission per sector = Emission j / Output j
ES = E.copy()
ES = E / xj[:, np.newaxis] # np.newaxis create a column vector from shape(4,) to shape(4,1)   
ES_df = pd.DataFrame(ES,
                    index   = ['Aluminium', 'Coal & Petroleum', 'Electricity'],
                    columns = ['Energy consumption [MJ/$]', 'CO2 emissions [kg/$]'])
ES_df

C:\Users\barsanti.INTRANET\AppData\Local\Temp\ipykernel_592\2543944024.py:6: FutureWarning: Support for multi-dimensional indexing (e.g. `obj[:, None]`) is deprecated and will be removed in a future version.  Convert to a numpy array before indexing instead.
  ES = E / xj[:, np.newaxis] # np.newaxis create a column vector from shape(4,) to shape(4,1)


Energy consumption [MJ/$]  CO2 emissions [kg/$]
Aluminium                      0.000000e+00          1.933896e+05
Coal & Petroleum               5.707513e+08          6.929249e+05
Electricity                    0.000000e+00          1.132931e+07

Since the energy consumption here only refers to the primary energy consumption, only the coal & petroleum sector actually use it, whilst the other sectors use secondary energy inputs. Nonetheless, in the transformation of primary to secondary energy, CO2 is emitted, which is why the CO2 emissions in the aluminium and electricity sectors are non null. 

#### EX1.4
*To be able to evaluate the embodied energy and the CO2 emissions, case specific data has to be collected. Calculate the vector of final demand.*

In [6]:
# -- Data

# demand matrix front panel
Y = np.array([[  5.9,   2.5], 
              [ 2.14,  0.32], 
              [ 15.2,  0.07],
              [ 30.4,  0.36]])
Y_df = pd.DataFrame(Y,
                    index   = ['Aluminium', 'Oil', 'Electricity', 'Gasoline during use'],
                    columns = ['Amount (kg or l or kWh/panel)', 'Price ($/kg or l or kWh)'])
Y_df

Amount (kg or l or kWh/panel)  Price ($/kg or l or kWh)
Aluminium                                     5.90                      2.50
Oil                                           2.14                      0.32
Electricity                                  15.20                      0.07
Gasoline during use                          30.40                      0.36

In [7]:
# -- Calculation

# calculate the final demand
Y_df.loc[:,'Final demand $'] = Y_df.loc[:,'Amount (kg or l or kWh/panel)'] * Y_df.loc[:,'Price ($/kg or l or kWh)']
Y_df.loc['Gasoline during use', 'Final demand $'] = 'not part of production process'
Y_df

Amount (kg or l or kWh/panel)  Price ($/kg or l or kWh)  \
Aluminium                                     5.90                      2.50   
Oil                                           2.14                      0.32   
Electricity                                  15.20                      0.07   
Gasoline during use                          30.40                      0.36   

                                     Final demand $  
Aluminium                                     14.75  
Oil                                          0.6848  
Electricity                                   1.064  
Gasoline during use  not part of production process

#### EX1.5
*Using the data provided in table 3, determine the necessary final output (in monetary units) in
each sector in order to manufacture one front panel. Identify the sector that generates the
most important contribution to the final output of the electricity sector.*

In [8]:
# -- Calculation

# Identity matrix
I = np.identity(len(A))

# final demand
Y = Y_df.loc[['Aluminium','Oil','Electricity'],'Final demand $'].to_numpy().copy()

# Total output needed to satisfy the demand of a front panel ( X = (I – A)^(- 1) * final demand )
X = np.dot(np.linalg.inv(I-A), Y)
X_df = pd.DataFrame(X,
                    index   = ['Aluminium', 'Oil', 'Electricity'],
                    columns = ['output per sector in $'])
X_df

output per sector in $
Aluminium                17.805178
Oil                       1.341465
Electricity               5.832196

In [9]:
# composition of the output from the electricity sector
X_el = np.multiply(np.linalg.inv(I-A)[2], Y)
X_el_df = pd.DataFrame(X_el,
                    index   = ['Aluminium', 'Oil', 'Electricity'],
                    columns = ['output from the el. sector $'])
X_el_df

output from the el. sector $
Aluminium                       4.758491
Oil                             0.008212
Electricity                     1.065493

Hence the largest input to the electricity sector comes from the aluminium sector, with 4.76$.

#### EX1.6
*Estimate the total non-renewable primary energy consumption induced by the manufacturing of a front panel.*

In [11]:
# Total non-renewable energy demand induced by the production of a front panel
# = Nonrenewable energy cons. per sector * Total output for a front panel per sector
FF = np.dot(X, ES[:,0])/10**6 # MJ
print(FF, ' MJ')

765.6426663392106  MJ
